Plunging Geodesics Kerr pull request

ConorDyson opened the pull request:
Pull request for plunging geodesics in Kerr, Analytic implementation in terms of Mino time, solves for generic plunges given {a,e,L,Q} and can specialise to plunges from the ISSO for parameterizations {a,r_{ISSO}} or {a,\theta_{inc}}. Units test have also been included, and packaged sits on top of existing ones. The only change to existing packages should be the addition of the line  "KerrGeodesics`KerrGeoPlunge`" in the KerrGeodesics.m run file.

ConorDyson committed:
Working Package-V1

Version1 of plunge package working for ISSO, Complex and Real Roots - ISSO currently in separate package to Generic solution file, neither implemented in the original GeoOrbit package yet.

ConorDyson committed:
Package working- needs to be made nice.

ConorDyson committed:
5/12/ update

ConorDyson committed:
Plots almost ready

ConorDyson committed:
Nice Plots

TestPlungesProducingPaperPlots

ConorDyson committed:
Incline, Radial ISSO param, Generic

Package working ISSO for radial param and inclined param. Commit also contain test file with flow and orbital plots, need to include initial conditions and return mino time crossing horizon still

ConorDyson committed:
Initial Conditions Working

ConorDyson committed:
Update: Plunges

Small bug fixed, now works for ICs. Noticed generic coord tIme solution still is not fully analytic Arctan term is causing issues needs to be massaged to an analytic form.

Commit:
More robust zero check.

ConorDyson committed:
Update:19/12/22

Commit:
for some reason this subpackage was not included in this development branch

Commit:
Proper loading of test Paclet. See https://bhptoolkit.org/mathematica-install-dev.html

Commit:
Implement an example of a proper error message. Also use Mathematica canonical keyword Automatic for empty input (instead of "Nan")

ConorDyson committed:
Merge branch 'Plunge' of https://github.com/ConorDyson/KerrGeodesics into Plunge

ConorDyson committed:
Update:20/12/22-Syntax&ICs

ConorDyson committed:
Analytic continuation fixed

MvdMeent committed:
Partial pass through code.

ConorDyson committed:
AngMomentum Fixed for ISSO case

ConorDyson committed:
Further Improvements

Real Root Code has some issues, added all crossing times and fixed format of complex roots file, also added all initial conditions properly.

ConorDyson committed:
Tasks Completed

All fixes seem to be working, just need to run a script to test all possible cases i.e. code works when it should and gives correct error messages when it shouldn't

ConorDyson committed:
Revert "Tasks Completed"

This reverts commit 8d1284f256105a45fe543252cb4075a88240be74.

ConorDyson committed:
Fixes for ISSO

L now works for all allowed r_I and ISSO works for equatorial case now.

ConorDyson committed:
Added Test File - Everything seems to work

ConorDyson committed:
UpdatedOutputs

MIno time and other outputs improved to correct form

ConorDyson committed:
Test FIle Improvement

Tests provide in proper Mathematica Test File now

ConorDyson committed:
File organisation

Moved test file to test folder and removed unnecessary additional files

Commit:
Message tests

ConorDyson committed:
Included unit tests as wlt files

ConorDyson committed:
Precision consistency and Test files implemented wlt's

ConorDyson committed:
Revert "Included unit tests as wlt files"

This reverts commit 1ad602dd31057145a4dee6f2345bf2c2a20e9c1f.

ConorDyson committed:
Revert "Precision consistency and Test files implemented wlt's"

This reverts commit 6f8e3e1d3347d836c56f18ae884efe3cb34a5e6b.

barrywardell commented on Kernel/KerrGeodesics.m:
```diff
@@ -1,10 +1,16 @@
 (* ::Package:: *)
 
+$BaseDirectory
```
Is this supposed to be here?

ConorDyson replied:
No its removed now, also seems as though a performed the pull request on a version which was a few days out of date in which the test files weren't completed yet, the correct form of the wlt should now be included.

barrywardell commented on Kernel/KerrGeoOrbit.m:
```diff
@@ -19,7 +19,7 @@
 
 
-(* ::Section::Closed:: *)
+(* ::Section:: *)
```
These changes to section open/closed status shouldn't be included. Similarly for the others in this file

barrywardell commented on Kernel/KerrGeoPlunge.m:
```diff
@@ -0,0 +1,544 @@
+	 "KerrGeodesics`FourVelocity`"}];
+
+
+KerrGeoPlunge::usage = "Takes either KerrGeoPlunge[a, Generic, En ,\[ScriptCapitalL],\[ScriptCapitalQ]] or KerrGeoPlunge[a, ISSO , RI] for generic Plunges or ISSO plunges respectively and returns a KerrGeoPlungeFunction[..] which stores the orbital trajectory and parameters. Here the ISSO plunges are paramaterised in terms of the choice of the radius of the ISSO for a given a there are range of a ...
```
We need to think carefully about this interface. Do we want separate functions for ISSO plunges? Or can everything be inferred automatically and unambiguously from context?

ConorDyson replied:
I've reworded this section slightly; the usage is as follows. 

For generic plunges you provide either KerrGeoPlunges[a, {e,L,Q}] or KerrGeoPlunges[a, {e,L,Q}, {t0,to,theta0,phi0}] if you want to include initial conditions

For ISSO, one provides either KerrGeoPlunges[a, "ISSORadialParam", RI] or  KerrGeoPlunges[a, "ISSOIncParam", theta_Inc] with IC's working the same as the generic case as a final option.

The package defaults to Generic unless the user explicitly inputs an ISSO option which I think should be sufficiently unambiguous but I'm happy to consider alternatives.

nielsw2 replied:
I agree the main issue will be deciding on the interface. This should probably go along with the discussion related to issue #37 and associated pull request #38. We were thinking of having default input parameters as (a,p,e,x) but then allowing associations to be passed if the user wishes to use alternative parameterizations.

We also need to decide if we want `KerrGeoPlunges` or to wrap this within `KerrGeoOrbit`

ConorDyson replied:
I'm not sure if parameterization by (a,p,e,x) is overly suitable for plunges. @MvdMeent also mentions in #37 how this parameterization breaks down for plunges. If we wanted a parameterization that works well for all classes of bound geodesics (including plunges), perhaps parameterizing by initial position and velocity, as Vojtech mentions in #37 might be best? Theres probably other reasons for why this might not be a good choice that I'm unaware of though.

If we do decide to wrap KerrGeoPlunges within KerrGeoOrbit, we do have to note the fact that for a large part of the parameter space, a given (a,E,L,Q) gives two solutions, one bound in the regular sense, and another "deeply bound"(plunging). So if we do this, we need to think about if we want the user to have to specify if they're looking for a plunging or bound geodesic when they run the code. Perhaps this isn't an issue if we can parameterize the plunges by (a,p,e,x) or initial positions and velocities, though?

I do think the ISSO case parameterization should be kept largely the same as it currently is however, because the nice thing about those solutions is being able to fully parameterize everything by a single choice of R_{ISSO}.

barrywardell reviewed (changes requested):
There appear to be multiple PlungesTest files: one in Kernel and two inside Tests/. Can these be merged into a single file? Also, can they be converted to wlt files for use with the automated testing system?

Further comments are also listed below.

ConorDyson committed:
Form with correct wlt file

nielsw2 commented:
I noticed that the a=0 case does not seem to work. e.g.,

`KerrGeoPlunge[0, {0.9, 0.7, 0.4}]` gives a bunch of errors.

ConorDyson commented:
The package currently doesn't work for the a=1 or a=0 case; I've added an error message that kills the run if either of these choices is made I'll see if the functions can be easily adapted to work in the a=0 case I'm pretty certain there is no easy way to make it work in the extremal case but ill check also.

ConorDyson committed:
Added a=0,1 error messages and removed Closed

ConorDyson committed:
Equations now working for a=0

Working for a=0 still gives error message for a=1. a=0 case has been extended through the manifold singularity but is still continuous and satisfies the EOM so seems to be working correctly.

ConorDyson committed:
working a=0

closed section opened

ConorDyson commented:
Both the generic case and ISSO now work for a=0.

nielsw2 commented:
Just so we don't forget. Following a meeting on the 24th of Feb. we decided on the following API:

```Mathematica
KerrGeoOrbit[a,"\[ScriptCapitalE]\[ScriptCapitalL]\[ScriptCapitalQ]"->{0.9,3,1.2},"InitialPhases"->{1,2,4}]
KerrGeoOrbit[a,"\[ScriptCapitalE]\[ScriptCapitalL]\[ScriptCapitalQ]"->{0.9,3,1.2},"OrbitType"->"Plunge"]
KerrGeoOrbit[a,"Subscript[r, ISSO]"->5,"OrbitType"->"Plunge"]
KerrGeoOrbit[a,"InitialPosition"->{1,2,3,4},"FourVelocity"->{1,5,6,7}]
```

This is the main change needed for before merging this pull request and #38

ConorDyson committed:
Added homoclinic orbit solutions

ConorDyson committed:
Resolved Merge Conflicts Between  Fork and Master
